### PR TITLE
fix issue#76

### PR DESCRIPTION
**Pull request by xwchris**
to fix this issue https://github.com/sysgears/webpack-virtual-modules/issues/76

**What's the problem this PR addresses?**
https://github.com/sysgears/webpack-virtual-modules/issues/76

**How did you fix it?**

change the fileTimestamp value for new version of webpack

### CODE REVIEW

**Review by gabeklein — APPROVED** [review]
Can confirm this fixed caching issue for me on Webpack version `>=5.28.0`!

**Comment by VincentGarreau** [comment]
Hope this gets merged soon :)

**Comment by larixer** [comment]
@xwchris Please add more comments to the code or provide clear explanation in the PR what is this code doing

File: src/index.ts [hunk 1]
```diff
@@ -4,6 +4,14 @@ import type { Compiler } from 'webpack';
...
 let inode = 45000000;
 
+let FS_ACCURACY = 2000;
+function applyMtime (mtime) {
+    if (FS_ACCURACY > 1 && mtime % 2 !== 0) FS_ACCURACY = 1;
+    else if (FS_ACCURACY > 10 && mtime % 20 !== 0) FS_ACCURACY = 10;
+    else if (FS_ACCURACY > 100 && mtime % 200 !== 0) FS_ACCURACY = 100;
+    else if (FS_ACCURACY > 1000 && mtime % 2000 !== 0) FS_ACCURACY = 1000;
+};
+
```
Comment (gabeklein) [inline src/index.ts:14]: This is an assumption but a quick google of "filesystem precision" shows that filesystems have differing amounts of timestamp precision. 

[Check this thread](https://unix.stackexchange.com/a/490990)
> The things I learned so far were:
> 
> - XFS and EXT3: second precision
> - EXT4: millisecond precision
> - NTFS: 100ns precision (OK, got this from the docs...)

So I figure this is addressed to that fact.

File: src/index.ts [hunk 2]
```diff
@@ -263,7 +271,14 @@ class VirtualModulesPlugin {
       const fts = compiler.fileTimestamps as any;
       if (virtualFiles && fts && typeof fts.set === 'function') {
         Object.keys(virtualFiles).forEach((file) => {
-          fts.set(file, +virtualFiles[file].stats.mtime);
+          const mtime = +virtualFiles[file].stats.mtime;
```
Comment (gabeklein) [inline src/index.ts:274]: The `+` in `+virtualFiles[file].stats.mtime` apparently casts "mtime" into a string. Savvy trick, didn't know that though `Number()` probably would've been clearer.

File: src/index.ts [hunk 3]
```diff
@@ -263,7 +271,14 @@ class VirtualModulesPlugin {
...
+          if (mtime) applyMtime(mtime);
+          // webpack 5.58.2
+          fts.set(file, {
+              accuracy: 0,
+              safeTime: mtime ? mtime + FS_ACCURACY : Infinity,
+              timestamp: mtime
+          });
```
Comment (gabeklein) [inline src/index.ts:281]: The [Webpack 5 watch graph](https://webpack.js.org/contribute/plugin-patterns/#monitoring-the-watch-graph) behavior has apparently changed as of Webpack 5 breaking this library which is the root of the issue preventing compatibility.

When this function is fired (a file is created or updated) the watch graph needs to be informed so it's included into the build in watch-mode. Webpack does this basically by comparing timestamps (makes sense). I haven't been able to find for-sure what `safeTime` is referring to, but it seems to be used here as "nearest time there is no update" or something like that.

**Review by gabeklein — APPROVED** [review]
@larixer - I've put down some comments describing what I think is happening. If @xwchris gets back to us great, but would love to see this change merged sooner than later!

Cheers

**Review by gabeklein — CHANGES_REQUESTED** [review]
Ok just compared source of `fileTimestamps` between Webpack 4 and Webpack 5. The signature changed from `number` to `FileSystemInfoEntry` and that's what's caused this compatibility issue.

That said, this PR is going to need some way to detect Webpack 4 and still use a number, otherwise it will cause a regression on non-webpack 5 builds.

File: src/index.ts (hunk 1 above)
Comment (larixer) [reply src/index.ts:14]: Yes, but it does not explain all the four values, so you are on the right way, but the explanation is still incomplete

File: src/index.ts (hunk 3 above)
Comment (larixer) [reply src/index.ts:281]: Good find, thanks!

**Comment by larixer** [comment]
What is PR also misses is tests, because the code has compatibility problems with Webpack 4.

**Comment by larixer** [comment]
Closing, the PR has no tests

**Pull request closed without merging**